File: alarm/src/domain/factories/SecurityRulesFactory.ts
import { TimeSlot } from '../core/rules/TimeSlot.js'
import { IntrusionRule } from '../core/rules/IntrusionRule.js'
import { RangeRule } from '../core/rules/RangeRule.js'
import { SecurityRuleId } from '../core/rules/SecurityRuleId.js'
import { Contact } from 'common/dist/domain/core/Contact.js'
import { Measure, ObjectClass } from '@common/domain/core'
import { v4 as uuidv4 } from 'uuid'

export class SecurityRulesFactory {
  static newId(): SecurityRuleId {
    return { value: uuidv4() }
  }

  static idOf(id: string): SecurityRuleId {
    return { value: id }
  }

  static intrusionRuleOf(
    id: SecurityRuleId,
    activeOn: string,
    creatorId: string,
    objectClass: ObjectClass,
    contacts: Contact[],
    description: string,
    validity: TimeSlot,
    enabled: boolean
  ): IntrusionRule {
    return {
      id,
      type: 'intrusion',
      activeOn,
      creatorId,
      objectClass,
      contacts,
      description,
      validity,
      enabled
    }
  }

  static createIntrusionRule(
    activeOn: string,
    creatorId: string,
    objectClass: ObjectClass,
    contacts: Contact[],
    description: string,
    validity: TimeSlot,
    enabled: boolean
  ): IntrusionRule {
    return this.intrusionRuleOf(
      this.newId(),
      activeOn,
      creatorId,
      objectClass,
      contacts,
      description,
      validity,
      enabled
    )
  }

  static rangeRuleOf(
    id: SecurityRuleId,
    activeOn: string,
    creatorId: string,
    contacts: Contact[],
    description: string,
    validity: TimeSlot,
    min: number,
    max: number,
    measure: Measure,
    enabled: boolean
  ): RangeRule {
    return {
      id: id,
      type: 'range',
      activeOn,
      creatorId,
      contacts,
      description,
      validity,
      min,
      max,
      measure,
      enabled
    }
  }

  static createRangeRule(
    activeOn: string,
    creatorId: string,
    contacts: Contact[],
    description: string,
    validity: TimeSlot,
    min: number,
    max: number,
    measure: Measure,
    enabled: boolean
  ): RangeRule {
    return this.rangeRuleOf(
      this.newId(),
      activeOn,
      creatorId,
      contacts,
      description,
      validity,
      min,
      max,
      measure,
      enabled
    )
  }

  static newTimeSlot(from: Date, to: Date): TimeSlot {
    if (from >= to) {
      throw new Error('Invalid time slot')
    }
    return { from, to }
  }
}

File: recognition/app/domain/factories/SecurityRulesFactory.py
import uuid
from datetime import datetime
from typing import List

from app.domain.core import ObjectClass
from app.domain.core.rules import SecurityRuleId, Contact, TimeSlot, IntrusionRule


class SecurityRulesFactory:

    @staticmethod
    def new_id() -> SecurityRuleId:
        return SecurityRuleId(value=str(uuid.uuid4()))

    @staticmethod
    def id_of(id_: str) -> SecurityRuleId:
        return SecurityRuleId(value=id_)

    @staticmethod
    def intrusion_rule_of(
        id_: SecurityRuleId,
        active_on: str,
        creator_id: str,
        object_class: ObjectClass,
        contacts: List[Contact],
        description: str,
        validity: TimeSlot,
        enabled: bool,
    ) -> IntrusionRule:
        return IntrusionRule(
            id=id_,
            type="intrusion",
            active_on=active_on,
            creator_id=creator_id,
            object_class=object_class,
            contacts=contacts,
            description=description,
            validity=validity,
            enabled=enabled,
        )

    @staticmethod
    def create_intrusion_rule(
        active_on: str,
        creator_id: str,
        object_class: ObjectClass,
        contacts: List[Contact],
        description: str,
        validity: TimeSlot,
        enabled: bool,
    ) -> IntrusionRule:
        return SecurityRulesFactory.intrusion_rule_of(
            SecurityRulesFactory.new_id(),
            active_on,
            creator_id,
            object_class,
            contacts,
            description,
            validity,
            enabled,
        )

    @staticmethod
    def new_time_slot(from_: datetime, to: datetime) -> TimeSlot:
        if from_ >= to:
            raise ValueError("Invalid time slot")
        return TimeSlot(from_=from_, to=to)
